Add timeline and followups to Changes and Problems

cconard96 opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #3222 #3223

Add timeline and followups to Changes and Problems, and drop tabs replaced by the timeline.

cconard96 commented:
I will need to install a clean master branch with my changes and check those issues. I was going to follow this PR up with adding followups for changes and problems per #3222. I'm thinking I may have accidentally used the localized class name in some cases for tables :(

trasher commented:
Should not followups be added on problems and changes before timeline? This sounds more logical to me.

cedric-anne commented:
On CSS, maybe renaming classes to somethink linke `ITIL*` would prevent duplication of classes and would prevent usage of dynamic variables to defines HTML class attributes.

```
.TicketValidation -> .ITILValidation
.TicketTask -> .ITILTask
.TicketContent -> .ITILContent
.ticket_title -> .ITIL_title
```

trasher commented:
I do not understand why you create new followup tables in this one; since you've opened another pull request for changes|problems followups integration.

Anyways, you have to fix unit tests, there are syntax issues ;)

cconard96 commented:
Apparently I already rebased this PR. I'll have to separate them locally and re-test.

I shouldn't of opened conflicting requests at the same time, and switching the bases seems to have made it worse. Can we focus on the followups pull request first since I've already changed it to be the base? I think I'm going to have a lot of code that can be removed from followups when they are added to the timeline.

trasher commented:
What about one PR only for both?

It will sure be a huge PR to review/handle, but I wonder if we can really separate them, without extra work (that would lend to issues finally) :/

Thoughts?

orthagh commented:
i agree with this proposal

cconard96 commented:
I agree. :)

trasher commented on front/change.form.php:
```diff
@@ -104,6 +104,19 @@
               //TRANS: %s is the user login
               sprintf(__('%s adds an actor'), $_SESSION["glpiname"]));
    Html::redirect(Change::getFormURLWithID($_POST['changes_id']));
+}  else if (isset($_REQUEST['delete_document'])) {
```
Not sure this change is relevant here

trasher replied:
OK :)

trasher commented:
I've tested your branch, this seems to work quite well. I saw 2 issues:
- Cannot refuse problem or change solution
- CSS issue on ticket followup on original description and followups

cconard96 commented:
Not seeing the CSS issue. Do you have a screenshot?
![screenshot from 2018-09-17 18-34-57](https://user-images.githubusercontent.com/17678637/45654138-6bc2ca00-baa8-11e8-84ca-f527f61d965e.png)

trasher commented:
I do not have screenshot, but background colors were missing. Will try again making sure my browser cache is clear; I'll keep you in touch.

trasher commented:
Just tested again: there is no bg on tickets timeline. CSS class is still `TicketFollowup`, that sounds wrong.

It seems some class may be factorized a bit (*followup); and I guess `CommonITILFollowup` may be renamed to `ITILFollowup`.

trasher commented on inc/change.class.php:
```diff
@@ -87,6 +87,11 @@ function canAssignToMe() {
    }
 
 
+   function canAddFollowups() {
```
This method is exactly the same for Changes and Problems (not for Tickets); so it should maybe be moved to CommonITILObject super class

trasher commented on inc/change.class.php:
```diff
@@ -159,6 +164,31 @@ function canCreateItem() {
+    *
+    * @return an array
+   **/
+   static function getReopenableStatusArray() {
```
This method is exactly the same for Changes, Problems and Tickets; so it must be moved to CommonITILObject super class

trasher commented on inc/change.class.php:
```diff
@@ -159,6 +164,31 @@ function canCreateItem() {
+      return Session::haveRight('followup', CREATE)
+             && in_array($this->fields["status"], $this->getClosedStatusArray())
+             && ($this->isAllowedStatus($this->fields['status'], self::INCOMING)
+                 || $this->isAllowedStatus($this->fields['status'], self::EVALUATION));
```
Not sure for `EVALUATION`, @orthagh?

trasher replied:
OK, fne, then Thanks

trasher commented on inc/change.class.php:
```diff
@@ -1163,4 +1210,39 @@ static function showListForItem(CommonDBTM $item) {
+    * will return an assoc array of users_id => array of roles.
+    * @return array[] of array[] of users and roles
+    * @since 0.90
+    * @deprecated
```
precise deprecated version

trasher commented on inc/change.class.php:
```diff
@@ -1163,4 +1210,39 @@ static function showListForItem(CommonDBTM $item) {
+    * @since 0.90
+    * @deprecated
+   **/
+   function getITILActors() {
```
I do not understrand why to add this method and deprecate it...

cconard96 replied:
Fixed. I also changed it so getTicketActors calls getITILActors instead of the other way around. It should make it a little easier to remove in the future.

trasher commented on inc/commonitilobject.class.php:
```diff
@@ -5675,14 +5672,868 @@ function getCalendar() {
+
+
+   /**
+    * @since 9.4.0
```
Please complete docblock

trasher commented on inc/commonitilobject.class.php:
```diff
@@ -5675,14 +5672,868 @@ function getCalendar() {
+
+
+   /**
+    * @since 9.4.0
```
Missing description and return type

trasher commented on inc/commonitilobject.class.php:
```diff
@@ -5675,14 +5672,868 @@ function getCalendar() {
+
+
+   /**
+    * @since 9.4.0
```
Missing description and return type

trasher commented on inc/commonitilobject.class.php:
```diff
@@ -5675,14 +5672,868 @@ function getCalendar() {
+
+
+   /**
+    * @since 9.4.0
```
Missing description and return type

trasher commented on inc/commonitilobject.class.php:
```diff
@@ -5675,14 +5672,868 @@ function getCalendar() {
+
+
+   /**
+    * @since 9.4.0
```
Missing description and return type, missing param doc and type

trasher commented on inc/commonitilobject.class.php:
```diff
@@ -1732,7 +1752,7 @@ private function addAdditionalActors($input) {
              && count($input['_additional_suppliers_assigns'])) {
 
-            $input2 = [
+                        $input2 = [
```
Fixed

trasher commented:
Hi,

I've added 2 commits:
1- to deprecate old method (must call `Toolbox::deprecate()`)
2- to factorize `getItilActors`.

trasher commented:
Regading ITILFollowup methods marked as deprecated; I guess they can be dropped (@orthagh do you agreee?)

By the way, same cleanup could be done for CommonITILTasks and ITILSolution

orthagh commented:
right checks seems to be bad also.
rightname property should be set in change and problem sub classes
and canXXX superflu methods cleaned.

currently some are based on 'task' rightname which is the ticket task right.
Change and problem don't have task sub right

trasher commented:
For tasks stuff, I've opened another ticket; this is not directly related (and current PR is probably  huge enough already :D ); see https://github.com/glpi-project/glpi/issues/4738

cconard96 commented:
I noticed some recent commits to master are in conflict with this PR, and some add back methods that were moved (like getSplittedSubmitButton). I havent had a chance to rebase the branch so that may be causing the issues. I cannot recreate them in my environment.

Thanks for the assistance @trasher.

trasher commented:
@cconard96 do not worry about conflicts right now. I'll do a rebase+ squash of the branch on my side, maybe tomorrow.
Just add commits for now if you have to.

trasher commented:
I've fixed private display (and used FA instead of images)
![screenshot_2018-10-05 glpi - tickets - 1146](https://user-images.githubusercontent.com/224733/46535642-650bc380-c8ac-11e8-81fe-383ffeb25314.png)

trasher commented:
Note: I was going to made display changes on master only from a separate PR, but code has changed here... I prefer not to waste time. Display fixes may be repoted on 9.3/bf branch :/

trasher commented:
I've rebased the branch against master, fixed conflits, and then squashed into two commits: one for all work on timeline, the other for the display of private tasks/followups (existing bug in 9.3).

For everyone: you'll have to drop your local `feature/itiltimeline` branch, then fetch remote, then grab the branch again.

cconard96 commented:
No problem. I enjoyed working on it.

cedric-anne commented on inc/itilfollowup.class.php:
```diff
@@ -54,94 +53,138 @@ class TicketFollowup  extends CommonDBTM {
     * @return boolean
    **/
-   function canCreateItem() {
+   function canUpdateITILItem() {
```
Is this function used ?

trasher replied:
Seems not :)

trasher commented on CHANGELOG.md:
```diff
@@ -22,6 +28,10 @@ The present file will list all changes made to the project; according to the
 - Remove `$CFG_GLPI['use_rich_text']` parameter. Will now be `true` per default.
 - Remove `$CFG_GLPI['ticket_timeline']` parameter. Will now be `true` per default.
 - Remove `$CFG_GLPI['ticket_timeline_keep_replaced_tabs']` parameter. Will now be `false` per default.
+- Move `TicketFollowup::showFormMassiveAction` to `ITILFollowup` and deprecate.
```
`show*` and massive no longer exists at all (and there is no replacement); no depreciation message will never be shown... Well do not see how to manage this nightmare

We've discussed about this "dead code"; it will be handled more globally in another PR;

cedric-anne replied:
I just put back empty TicketFollowup class to handle deprecation state.

cedric-anne commented:
@orthagh Should we update default notifications of changes and problems to add followups (copy/paste from ticket) ?

cedric-anne commented on inc/api.class.php:
```diff
@@ -2280,7 +2280,6 @@ static function getHatoasClasses($itemtype) {
       //specific case
       switch ($itemtype) {
          case 'Ticket' :
-            $hclasses[] = "TicketFollowup";
```
As we discuss together, response is yes.

cedric-anne commented:
Hi,

I made some changes:
 - fix some side effects,
 - fix migration (existing followups `itemtype` was not set to `Ticket`),
 - factorize some code,
 - handle deprecation of `TicketFollowup` class,
 - handle `ITILFollowup` in API,
 - handle followups in Changes and Problems notifications.

Regards

trasher commented on inc/notificationtargetcommonitilobject.class.php:
```diff
@@ -525,6 +560,85 @@ function addTaskAssignGroup($options = []) {
+
+
+   /**
+    * @param $data   array
```
Either doc is missing or should be entirely dropped

trasher commented on inc/notificationtargetcommonitilobject.class.php:
```diff
@@ -525,6 +560,85 @@ function addTaskAssignGroup($options = []) {
+
+   /**
+    * Restrict by profile and by config
+    * to avoid send notification to a user without rights
```
Missing return type

trasher commented on inc/notificationtargetticket.class.php:
```diff
@@ -39,55 +39,18 @@
-         }
-
-         // Private object and no right to see private items : do not send
-         if ($this->isPrivate()
```
Is this one missing? Or implemented another way?

cedric-anne replied:
This parts has been moved to parent class, as it will be used for Changes and Problems. https://github.com/glpi-project/glpi/pull/4540/files/395de44ee6de273791687bfedc15834e48686b50#diff-ea1db5d4dfd69e74f2c6f17821279f2bR80

trasher replied:
Ok :)
